cmd/launcher/flare.go

```diff
@@ -0,0 +1,331 @@
+
+		// not documented via flags on purpose
+		enrollSecret    = env.String("KOLIDE_LAUNCHER_ENROLL_SECRET", "flare_ping")
+		serverURL       = env.String("KOLIDE_LAUNCHER_HOSTNAME", *flHostname)
+		insecureTLS     = env.Bool("KOLIDE_LAUNCHER_INSECURE", false)
+		insecureGRPC    = env.Bool("KOLIDE_LAUNCHER_INSECURE_GRPC", false)
+		flareSocketPath = env.String("FLARE_SOCKET_PATH", filepath.Join(os.TempDir(), "flare.sock"))
+
+		certPins [][]byte
+		rootPool *x509.CertPool
+	)
+	flagset.Usage = commandUsage(flagset, "launcher flare")
+	if err := flagset.Parse(args); err != nil {
+		return err
+	}
+
+	id := ulid.New()
+	b := new(bytes.Buffer)
+	reportName := fmt.Sprintf("kolide_launcher_flare_report_%s", id)
+	tarOut, err := os.Create(fmt.Sprintf("%s.tar.gz", reportName))
+	if err != nil {
+		fatal(b, err)
+	}
+	defer func() {
+		if err := tarOut.Close(); err != nil {
+			fatal(b, err)
+		}
+	}()
+
+	tw := tar.NewWriter(tarOut)
+
+	// create directory at root of tar file
+	baseDir := filepath.ToSlash(reportName)
+	hdr := &tar.Header{
+		Name:     baseDir + "/",
+		Mode:     0755,
+		ModTime:  time.Now().UTC(),
+		Typeflag: tar.TypeDir,
+	}
+
+	if err := tw.WriteHeader(hdr); err != nil {
+		fatal(b, err)
+	}
+
+	defer func() {
+		hdr := &tar.Header{
+			Name: filepath.Join(baseDir, fmt.Sprintf("%s.log", id)),
+			Mode: int64(os.ModePerm),
+			Size: int64(b.Len()),
+		}
+
+		if err := tw.WriteHeader(hdr); err != nil {
+			fatal(b, err)
+		}
+
+		if _, err := tw.Write(b.Bytes()); err != nil {
+			fatal(b, err)
+		}
+
+		if err := tw.Close(); err != nil {
+			fatal(b, err)
+		}
+	}()
+
+	output(b, stdout, "Starting Launcher Diagnostics\n")
+	output(b, stdout, "ID: %s\n", id)
+	user, err := user.Current()
+	if err != nil {
+		fatal(b, err)
+	}
+	output(b, stdout, "CurrentUser: %s uid: %s\n", user.Username, user.Uid)
+	v := version.Version()
+	jsonVersion, err := json.Marshal(&v)
+	if err != nil {
+		fatal(b, err)
+	}
+	output(b, stdout, "%v\n", string(jsonVersion))
+
+	logger := log.NewLogfmtLogger(b)
+	err = reportGRPCNetwork(
+		logger,
+		serverURL,
+		insecureTLS,
+		insecureGRPC,
+		enrollSecret,
+		certPins,
+		rootPool,
+	)
+	output(b, stdout, "GRPC Connection ...%v\n", err == nil)
+	if err != nil {
+		output(b, fileOnly, "reportGRPCNetwork error: %s", err)
+	}
+
+	err = reportOsqueryProcessInfo(logger, flareSocketPath, b)
+	if err != nil {
+		output(b, fileOnly, "reportOsqueryProcessInfo error: %s", err)
+	}
+	output(b, stdout, "Osqueryi ProcessInfo ...%v\n", err == nil)
+
+	err = reportNotaryPing(logger)
+	if err != nil {
+		output(b, fileOnly, "reportNotaryPing error: %s", err)
+	}
+	output(b, stdout, "Osqueryi Ping Notary ...%v\n", err == nil)
+
+	return nil
+}
+
+type outputDestination int
+
+const (
+	fileOnly outputDestination = iota
+	stdout
+)
+
+func fatal(w io.Writer, err error) {
+	output(w, stdout, "error: %s\n", err)
+	os.Exit(1)
+}
+
+func output(w io.Writer, printTo outputDestination, f string, a ...interface{}) error {
```

Comment:
What is printTo if not "stdout" ?

Reply:
fileOnly. 

The way this output helper works is it lets me decide if I want to send some output to the users terminal, or if I want to silently write it to the log file in the report. I want to capture everything in the log file, but the stdout output should be a bare minimum.

Reply:
Ah. I'd reasonable about that as a boolean. But that makes sense

Reply:
boolean could work, I was thinking there might be a third case down the road, but I also like functions that look like this:

`foo(true, true, false, true)`. It's hard to read and know the meaning of these booleans. An enum can be more descriptive.